modin/test/exchange/dataframe_protocol/omnisci/test_protocol.py

```diff
@@ -0,0 +1,259 @@
+    chunk_groups = [1, 2, 7]
+    chunk_col_ilocs = [
+        slice(
+            i * len(pd_df.columns) // len(chunk_groups),
+            (i + 1) * len(pd_df.columns) // len(chunk_groups),
+        )
+        for i in range(len(chunk_groups))
+    ]
+
+    pd_chunk_groups = [
+        split_df_into_chunks(pd_df.iloc[:, cols], n_chunks)
+        for n_chunks, cols in zip(chunk_groups, chunk_col_ilocs)
+    ]
+    at_chunk_groups = [
+        pa.concat_tables([pa.Table.from_pandas(pd_df) for pd_df in chunk_group])
+        for chunk_group in pd_chunk_groups
+    ]
+
+    chunked_at = at_chunk_groups[0]
+    # TODO: appending columns one by one looks inefficient, is there a better way?
+    for _at in at_chunk_groups[1:]:
+        for field in _at.schema:
+            chunked_at = chunked_at.append_column(field, _at[field.name])
+    md_df = from_arrow(chunked_at)
+
+    # verify that test generated the correct chunking
+    internal_at = md_df._query_compiler._modin_frame._partitions[0][0].get()
+    for n_chunks_group, cols in zip(chunk_groups, chunk_col_ilocs):
+        for col in internal_at.select(range(cols.start, cols.stop)).columns:
+            assert len(col.chunks) == n_chunks_group
+
+    n_chunks = md_df.__dataframe__().num_chunks()
+
+    exported_df = export_frame(md_df)
+    df_equals(md_df, exported_df)
+
+    exported_df = export_frame(md_df, n_chunks=n_chunks)
+    df_equals(md_df, exported_df)
+
+    exported_df = export_frame(md_df, n_chunks=n_chunks * 2)
+    df_equals(md_df, exported_df)
+
+    exported_df = export_frame(md_df, n_chunks=n_chunks * 3)
+    df_equals(md_df, exported_df)
+
+
+def test_export_when_delayed_computations():
+    """
+    Test that export works properly when OmnisciOnNative has delayed computations.
+
+    If there are delayed functions and export is required, it has to trigger the execution
+    first prior materializing protocol's buffers, so the buffers contain actual result
+    of the computations.
+    """
+    # OmniSci can't import 'uint64' as well as booleans, so exclude them
+    # issue for bool: https://github.com/modin-project/modin/issues/4299
+    data = get_data_of_all_types(has_nulls=True, exclude_dtypes=["uint64", "bool"])
+    md_df = pd.DataFrame(data)
+    pd_df = pandas.DataFrame(data)
+
+    md_res = md_df.fillna({"float32_null": 32.0, "float64_null": 64.0})
+    pd_res = pd_df.fillna({"float32_null": 32.0, "float64_null": 64.0})
+    assert (
+        not md_res._query_compiler._modin_frame._has_arrow_table()
+    ), "There are no delayed computations for the frame"
+
+    exported_df = export_frame(md_res)
+    df_equals(exported_df, pd_res)
+
+
+@pytest.mark.parametrize("data_has_nulls", [True, False])
+def test_simple_import(data_has_nulls):
+    """Test that ``modin.pandas.utils.from_dataframe`` works properly."""
+    data = get_data_of_all_types(data_has_nulls)
+
+    md_df_producer = pd.DataFrame(data)
+    md_df_consumer = from_dataframe(md_df_producer)
```

Comment:
won't this short-circuit, i.e. return back `md_df_producer` untouched?

Reply:
added assertions verifying that it doesn't
```python
    # TODO: the following assertions verify that `from_dataframe` doesn't return
    # the same object untouched due to optimization branching, it actually should
    # do so but the logic is not implemented yet, so the assertions are passing
    # for now. It's required to replace the producer's type with a different one
    # to consumer when we have some other implementation of the protocol as the
    # assertions may start failing shortly.
    assert md_df_producer is not md_df_consumer
    assert (
        md_df_producer._query_compiler._modin_frame
        is not md_df_consumer._query_compiler._modin_frame
    )
```